File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):

        Set = set()
        j = 0
        maxlength = 0

        for i in range(len(s)):

            if s[i] in Set:

                length = i - j
                maxlength = max(maxlength, length)

                while s[j] != s[i]:
                    Set.remove(s[j])
                    j += 1

                Set.remove(s[j])
                j += 1

            Set.add(s[i])

        length = len(s) - j
        maxlength = max(maxlength, length)

        return maxlength

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int maxlength = 0;
        int j = 0;

        for(int i=0; i<s.length(); i++){
            if(set.contains(s.charAt(i))){
                maxlength = Math.max(maxlength, i-j);
                while(s.charAt(j) != s.charAt(i)){
                    set.remove(s.charAt(j));
                    j++;
                }
                j++;
                set.add(s.charAt(i));
            }
            set.add(s.charAt(i));
        }
        maxlength = Math.max(maxlength, s.length()-j);
        return maxlength;
    }
}